src/firefly/java/edu/caltech/ipac/firefly/server/security/SsoCheckFilter.java

```diff
@@ -59,26 +70,50 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
                     // token has expires on our end's.. should refresh it.
                     authToken = SsoAdapter.getAdapter().refreshAuthToken(authToken);
                 }
-                if (authToken == null && authRequired) {
-                    if (getHeader(req, REQUEST_WITH, "").equals(AJAX_REQUEST)) {
-                        // has not authenticated or session has expired..
-                        // but, since ajax request will not respond to redirect.  we'll respond with a 401-Unauthorized
-                        res.sendError(401);
-                        return;
-                    } else {
-                        // check user auth..
-                        String qstr = req.getQueryString() == null ? "" : "?" + req.getQueryString();
-                        res.sendRedirect(SsoAdapter.getAdapter().makeAuthCheckUrl(
-                                ServerContext.getRequestOwner().getRequestAgent().getRequestUrl() + qstr));
-                        // after auth check, request will resume at SsoVerifyServlet.  no need to process this request.
-                        return;
+                if (authRequired) {
+                    if (authToken == null) {
+                        if (getHeader(req, REQUEST_WITH, "").equals(AJAX_REQUEST)) {
+                            // has not authenticated or session has expired..
+                            // but, since ajax request will not respond to redirect.  we'll respond with a 401-Unauthorized
+                            res.setHeader("WWW-Authenticate", "custom realm=\"apps\"");
+                            res.sendError(401);
+                            return;
+                        } else {
+                            // check user auth..
+                            String qstr = req.getQueryString() == null ? "" : "?" + req.getQueryString();
+                            res.sendRedirect(SsoAdapter.getAdapter().makeAuthCheckUrl(
+                                    ServerContext.getRequestOwner().getRequestAgent().getRequestUrl() + qstr));
+                            // after auth check, request will resume at SsoVerifyServlet.  no need to process this request.
+                            return;
+                        }
+                    }
+                    // is authenticated.. now check to see if user is allowed access
+                    if (authRequired && allowAccess != null && allowAccess.size() > 0) {
```

Comment:
Just curious how this Filter servlet is triggered and who is setting "allow" parameter? Can we be sure that allowAccess is non-null when authRequired is true?